src/syn/io/der.rs

```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+            Length::Indefinite => Some(self.buffer.read_position..self.buffer.write_position),
+            _ => None
+        };
+
+        if let Some(ref range1) = range {
+            if self.buffer.byte_len()*8 < range1.end {
+                return Err(Error::EndOfStream);
+            }
+        }
+
+        f(self)
     }
 
     #[inline]
     fn read_sequence_of<C: sequenceof::Constraint, T: ReadableType>(
         &mut self,
     ) -> Result<Vec<T::Type>, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[sequence of] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 16 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        let mut last_read_position = self.buffer.read_position;
+
+        if let Length::Definite(l) = &length {
+            last_read_position += l;
+        } else {
+            return Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
+
+        let mut vec = Vec::new();
+        while self.buffer.read_position < last_read_position {
+            vec.push(T::read_value(self)?);
+        }
+        Ok(vec)
+    }
+
+    fn read_set<C: set::Constraint, S: Sized, F: Fn(&mut Self) -> Result<S, Self::Error>>(&mut self, f: F) -> Result<S, Self::Error> {
+        self.read_sequence::<C, S, F>(f)
+    }
+
+    fn read_set_of<C: setof::Constraint, T: ReadableType>(&mut self) -> Result<Vec<<T as ReadableType>::Type>, Self::Error> {
+        self.read_sequence_of::<C, T>()
     }
 
     #[inline]
     fn read_enumerated<C: enumerated::Constraint>(&mut self) -> Result<C, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[enumerated] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Length::Definite(l) = length {
+            let index = self.buffer.read_i64_number(l)? as u64;
+            C::from_choice_index(index).ok_or(Error::InvalidChoiceIndex(index, C::STD_VARIANT_COUNT))
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_choice<C: choice::Constraint>(&mut self) -> Result<C, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[choice] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        let index = tag as u64;
+
+        if let Class::ContextSpecific = &class {
+        } else {
+            return Err(Error::InvalidType)
+        }
```

Comment:
(With `asn1rs::model::Type`) this can be collapsed to
```rust
if !matches!(tag, Tag::ContextSpecific(_)) {
    return Err(Error::InvalidType);
}
```